### PR TITLE
Fix #11421 computation of Set+2

### LINKED ISSUE #11421: Set+2 should not equal Set+1
ie this should fail:
~~~coq
Definition plus1_plus1 : Type@{Set+1} := Type@{Set+1}.
~~~

I haven't produced a proof of False from this but there probably is one.

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fixes #11421

### CODE REVIEW

**Comment by JasonGross** [comment]
Deserves a bug report and an entry in the critical bugs list?

Also should this target 8.11?

Also, nice catch!

**Comment by gares** [comment]
Wouldn't it make sense to change the signature of is_small?

**Comment by SkySkimmer** [comment]
In what way?

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
I think that this does need a critical entry before merging. @SkySkimmer can you write one quickly? The fix is small so it shouldn't delay the 8.11 release, which is close to imminent.

**Comment by gares** [comment]
> In what way?

The bug seems that the check "is_small" is used to detect the sort set (and others I guess) but here the sort is represented by a pair, and the bug was that the second component was disregarded. Maybe is_small should take the pair (eg always say no if the input is a successor of a small sort).

I'm not familiar with universe code, but the bug is a striking example of data not fitting the API.

**Comment by SkySkimmer** [comment]
>The bug seems that the check "is_small" is used to detect the sort set (and others I guess) but here the sort is represented by a pair, and the bug was that the second component was disregarded. Maybe is_small should take the pair (eg always say no if the input is a successor of a small sort).

That would be horrible to use when working with raw levels.

File: kernel/univ.ml [hunk 1]
```diff
@@ -345,8 +345,8 @@ struct
           (Level.is_prop u && not (Level.is_sprop v))
         else false
 
-    let successor (u,n) =
-      if Level.is_small u then type1
+    let successor (u,n as e) =
+      if is_small e then type1
```
Comment (SkySkimmer) [inline kernel/univ.ml]: @gares maybe you will like this diff better than the previous one (this `is_small` is `Expr.is_small` from the current module)

**Comment by SkySkimmer** [comment]
>I think that this does need a critical entry before merging. @SkySkimmer can you write one quickly? The fix is small so it shouldn't delay the 8.11 release, which is close to imminent.

added

File: dev/doc/critical-bugs [hunk 2]
```diff
@@ -167,6 +167,19 @@ Universes
   GH issue number: none
   risk: unlikely to be triggered in interactive mode, not present in batch mode (i.e. coqc)
 
+  component: algebraic universes
+  summary: Set+2 was incorrectly simplified to Set+1
+  introduced: V8.10 (with the SProp commit 75508769762372043387c67a9abe94e8f940e80a)
+  impacted released: V8.10.0 V8.10.1 V8.10.2
```
Comment (Zimmi48) [inline dev/doc/critical-bugs]: It's supposed to be "impacted released versions" which could be shortened to "impacted releases" but not "impacted released". The same error appears in the entry just above, from which you probably copied the template.

```suggestion
  impacted released versions: V8.10.0 V8.10.1 V8.10.2
```

**Comment by Zimmi48** [comment]
FTR the dual use of the "priority: blocker" and "priority: critical" labels achieves to convince me that "priority: critical" has a status that is not exactly the same as the other "priority" labels. Therefore, I hereby announce that I'm going to rename the "priority: critical" label into "kind: critical bug" (could also be "kind: soundness issue" but it would be less dramatic). If anybody disagrees, this kind of change is easy to undo.

**Comment by SkySkimmer** [comment]
Perhaps this is a good time to stop using "critical" and replace it with something more descriptive like "inconsistency" too.

**Comment by Zimmi48** [comment]
>Perhaps this is a good time to stop using "critical" and replace it with something more descriptive like "inconsistency" too.

Feel free to rename further.

**Comment by Zimmi48** [comment]
I don't want to be a bother but this would deserve a changelog entry too. Feel free to merge this before it gets one though, as we can also add it directly into #11372.

**Comment by ppedrot** [comment]
@SkySkimmer Can you rebase to get a clean CI?

**Comment by ppedrot** [comment]
@SkySkimmer can you also fix @Zimmi48's minor comments?

File: dev/doc/critical-bugs (hunk 2 above)
Comment (JasonGross) [inline dev/doc/critical-bugs]: It does not seem to be present in 8.10.2; with jscoq, I get
```
Definition plus1_plus1 : Type@{Set+1} := Type@{Set+1}.
(* Syntax error: [universe] expected after '@{' (in [constr:sort]). *)
```

**Comment by SkySkimmer** [comment]
try

Polymorphic Definition foo@{u} := Type@{u+1}.

Definition bar : foo@{Set}.
Proof. exact_no_check foo@{Set}. Qed.


Gaëtan Gilbert

**Comment by JasonGross** [comment]
> Polymorphic Definition foo@{u} := Type@{u+1}.
> Definition bar : foo@{Set}.
> Proof. exact_no_check foo@{Set}. Qed.

The `Qed` gives an error: "The term "foo@{Set}" has type "Type@{Set+2}" while it is expected to have type "foo@{Set}"."

**Comment by SkySkimmer** [comment]
how about

Polymorphic Definition foo@{u} := Type@{u+1}.
Definition s1 := Eval cbv in foo@{Set}.
Check s1 : s1.

Gaëtan Gilbert

**Comment by JasonGross** [comment]
Indeed, that works.  Here's your proof of `False`.  Note that we need to use `exact_no_check` in some places to get around the error message "Algebraic universe on the right".
```coq
Require Import Coq.Logic.Hurkens.

Section Paradox.

(** ** Universe [U] is equal to one of its elements. *)

Polymorphic Definition plus1@{u} := Type@{u+1}.
Let Set1 := Eval cbv in plus1@{Set}.
Let U : Type := Set1.
Let A:U := Set1.
Let h : U=A := eq_refl Set1.

(** ** Universe [U] is a retract of [A] *)

(** The following context is actually sufficient for the paradox to
    hold. The hypothesis [h:U=A] is only used to define [down], [up]
    and [up_down]. *)

Let down (X:U) : A := @eq_rect _ _ (fun X => X) X _ h.
Let up   (X:A) : U := @eq_rect_r _ _ (fun X => X) X _ h.

Lemma up_down : forall (X:U), up (down X) = X.
Proof.
  unfold up,down.
  rewrite <- h.
  reflexivity.
Qed.

Theorem paradox : False.
Proof.
  clearbody A h.
  Generic.paradox p.
  (** Large universe *)
  + exact U.
  + exact (fun X=>X).
  + cbn. exact (fun X F => forall x:X, F x).
  + cbn. exact (fun _ _ x => x).
  + cbn. exact (fun _ _ x => x).
  + exact (fun F => forall x:A, F (up x)).
  + cbn. exact (fun _ f => fun x:A => f (up x)).
  + cbn. intros * f X.
    specialize (f (down X)).
    rewrite up_down in f.
    exact f.
  (** Small universe *)
  + exact A.
  (** The interpretation of [A] as a universe is [U]. *)
  + cbn. exact up.
  + cbn. refine (fun u F => down _). cbv. exact_no_check (forall x, up (F x)). (* no_check to get around "algebraic universe on the right *)
  + cbn. refine (fun u F => down _). cbv. exact_no_check (forall x, up (F x)). (* no_check to get around "algebraic universe on the right *)
  + cbn. exact (down False).
  + rewrite up_down in p.
    exact p.
  + cbn. easy.
  + cbn. intros ? f X.
    destruct (up_down X). cbn.
    reflexivity.
  + cbn. intros ? ? f.
    rewrite up_down.
    exact f.
  + cbn. intros ? ? f.
    rewrite up_down in f.
    exact f.
  + cbn. intros ? ? f.
    rewrite up_down.
    exact f.
  + cbn. intros ? ? f.
    rewrite up_down in f.
    exact f.
Qed.

End Paradox.

Check paradox.
Print Assumptions paradox.
```

File: dev/doc/critical-bugs (hunk 2 above)
Comment (ppedrot) [reply dev/doc/critical-bugs]: @SkySkimmer you still need to fix this

Comment (SkySkimmer) [reply dev/doc/critical-bugs]: What's there to fix?

Comment (ppedrot) [reply dev/doc/critical-bugs]: Took care of it finally, this will be faster.

**Comment by ppedrot** [comment]
Should be ready now after the typo fix, will merge.

**Pull request merged**